src/prefect/cli/cloud.py

```diff
@@ -35,110 +49,234 @@
 app.add_typer(cloud_app)
 
 
-def build_url_from_workspace(workspace: Dict) -> str:
-    return (
-        f"{PREFECT_CLOUD_API_URL.value()}"
-        f"/accounts/{workspace['account_id']}"
-        f"/workspaces/{workspace['workspace_id']}"
-    )
+# Set up a little API server for browser based `prefect cloud login`
+
+
+def set_login_api_ready_event():
+    login_api.extra["ready-event"].set()
+
+
+login_api = FastAPI(on_startup=[set_login_api_ready_event])
+
+login_api.add_middleware(
+    CORSMiddleware,
+    allow_origins=["*"],
+    allow_methods=["*"],
+    allow_headers=["*"],
+)
+
+
+class LoginSuccess(BaseModel):
+    api_key: str
+
+
+class LoginFailed(BaseModel):
+    reason: str
+
+
+class LoginResult(BaseModel):
+    type: Literal["success", "failure"]
+    content: Union[LoginSuccess, LoginFailed]
+
+
+class ServerExit(Exception):
+    pass
+
+
+@login_api.post("/success")
+def receive_login(payload: LoginSuccess):
+    login_api.extra["result"] = LoginResult(type="success", content=payload)
+    login_api.extra["result-event"].set()
+
+
+@login_api.post("/failure")
+def receive_failure(payload: LoginFailed):
+    login_api.extra["result"] = LoginResult(type="failure", content=payload)
+    login_api.extra["result-event"].set()
+
+
+async def serve_login_api(cancel_scope):
+    config = uvicorn.Config(login_api, port=3001, log_level="critical")
```

Comment:
```suggestion
    config = uvicorn.Config(login_api, port=0, log_level="critical")
```
I haven't tested this but in theory it should work